codegen/go_transform.go

```diff
@@ -0,0 +1,652 @@
+package codegen
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"goa.design/goa/expr"
+)
+
+var (
+	// TransformGoArrayT is the template to generate Go array transformation
+	// code.
+	TransformGoArrayT *template.Template
+	// TransformGoMapT is the template to generate Go map transformation
+	// code.
+	TransformGoMapT *template.Template
+)
+
+// NOTE: can't initialize inline because https://github.com/golang/go/issues/1817
+func init() {
+	TransformGoArrayT = template.Must(template.New("transformGoArray").Funcs(template.FuncMap{
+		"transformAttribute": transformAttributeHelper,
+		"loopVar":            arrayLoopVar,
+	}).Parse(transformGoArrayTmpl))
+	TransformGoMapT = template.Must(template.New("transformGoMap").Funcs(template.FuncMap{
+		"transformAttribute": transformAttributeHelper,
+		"loopVar":            mapLoopVar,
+	}).Parse(transformGoMapTmpl))
+}
+
+type (
+	// GoTransformer defines the fields to generate Go code when transforming
```

Comment:
GoTransformer is a Transformer that generates Go code for converting a struct represented as an attribute expression into a different struct also represented as a an attribute expression.

(please explain what the rules for the conversions are at a high level and give an example of whether this is used).